### PR TITLE
Security JPA: support Hibernate multitenancy

### LINKED ISSUE #16700: JPA security : allow @UserDefinition to work with multitenant datasource
## Description

Hi, posting as enhancement but this may be a bug or missing feature.
We're are using quarkus.hibernate-orm.multitenant=DATABASE. Our custom TenantResolver or TenantConnectionResolver implementations work fine, in a @RequestScoped resolving tenant

However, we're facing a problem with JPA, https://quarkus.io/guides/security-jpa . A `io.quarkus.security.jpa.UserDefinition` is needed to trigger the `io.quarkus.security.jpa.runtime.JpaIdentityProvider`

The problem is that it will fail with the following at startup, because TenantConnectionResolver will not be bound. So the @UserDefinition does not understand how to get the tenant.

Is there a technique to set the tenant in the use case of a secured JPA @UserDefinition ?

```
    org.hibernate.HibernateException: SessionFactory configured for multi-tenancy, but no tenant identifier specified
        at org.hibernate.internal.AbstractSharedSessionContract.<init>(AbstractSharedSessionContract.java:172)
        at org.hibernate.internal.AbstractSessionImpl.<init>(AbstractSessionImpl.java:29)
        at org.hibernate.internal.SessionImpl.<init>(SessionImpl.java:221)
    ....
        at io.quarkus.security.jpa.runtime.JpaIdentityProvider$1.get(JpaIdentityProvider.java:35)
        at io.quarkus.security.runtime.QuarkusIdentityProviderManagerImpl$1$1$1$1.run(QuarkusIdentityProviderManagerImpl.java:58)

```
## Implementation ideas

To reproduce the exception, set quarkus.hibernate-orm.multitenant=DATABASE and annotate a single entity with @UserDefinition as per https://quarkus.io/guides/security-jpa

### PR DESCRIPTION
**Pull request by michalvavrik**
fixes: #16700

Makes sure the CDI request scope is activated when Hibernate multitenancy is enabled. The requirement on active request scope comes down to this:

https://github.com/quarkusio/quarkus/blob/5afc0ec8a525e3f632aebd9ff4401d87b1f5b983/extensions/hibernate-orm/runtime/src/main/java/io/quarkus/hibernate/orm/runtime/tenant/HibernateCurrentTenantIdentifierResolver.java#L32

### CODE REVIEW

File: extensions/security-jpa/deployment/src/main/java/io/quarkus/security/jpa/deployment/QuarkusSecurityJpaProcessor.java [hunk 1]
```diff
@@ -234,6 +250,32 @@ private ResultHandle lookupUserById(JpaSecurityDefinition jpaSecurityDefinition,
...
+    }
+
+    private static boolean shouldActivateCDIReqCtx(HibernateOrmConfig hibernateOrmConfig,
+            SecurityJpaBuildTimeConfig secJpaConfig) {
+        var puConfig = hibernateOrmConfig.getAllPersistenceUnitConfigsAsMap().get(secJpaConfig.persistenceUnitName());
+        if (puConfig == null) {
+            throw new ConfigurationException("Persistence unit '" + secJpaConfig.persistenceUnitName()
+                    + "' specified with the 'quarkus.security-jpa.persistence-unit-name' configuration property"
+                    + " does not exist. Please set valid persistence unit name.");
+        }
```
Comment (yrodiere) [inline extensions/security-jpa/deployment/src/main/java/io/quarkus/security/jpa/deployment/QuarkusSecurityJpaProcessor.java:268]: Note it's technically possible for the default persistence unit to not have any configuration, in which case the map wouldn't return any entry.

Are you sure you want to rely on `HibernateOrmConfig` directly? It seems dodgy. Wouldn't it make more sense to consume `io.quarkus.hibernate.orm.deployment.PersistenceUnitDescriptorBuildItem`? It has the advantage of being present regardless of configuration quirks, and exposes the multitenancy strategy through `getConfig().getMultiTenancyStrategy()`.

File: extensions/security-jpa/deployment/src/test/resources/multitenant-persistence-unit/application.properties [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+quarkus.datasource.one.db-kind=h2
+quarkus.datasource.one.username=sa
+quarkus.datasource.one.password=sa
+quarkus.datasource.one.jdbc.url=jdbc:h2:mem:one
+
+quarkus.datasource.two.db-kind=h2
+quarkus.datasource.two.username=sa
+quarkus.datasource.two.password=sa
+quarkus.datasource.two.jdbc.url=jdbc:h2:mem:one
```
Comment (yrodiere) [inline extensions/security-jpa/deployment/src/test/resources/multitenant-persistence-unit/application.properties:14]: Shouldn't this be:

```suggestion
quarkus.datasource.two.jdbc.url=jdbc:h2:mem:two
```

File: extensions/security-jpa/runtime/src/main/java/io/quarkus/security/jpa/runtime/JpaIdentityProvider.java [hunk 3]
```diff
@@ -37,16 +39,16 @@ public Uni<SecurityIdentity> authenticate(UsernamePasswordAuthenticationRequest
...
             @Override
             public SecurityIdentity get() {
-                EntityManager em = entityManagerFactory.createEntityManager();
-                ((org.hibernate.Session) em).setHibernateFlushMode(FlushMode.MANUAL);
-                ((org.hibernate.Session) em).setDefaultReadOnly(true);
-                try {
-                    return authenticate(em, request);
+                if (requireActiveCDIRequestContext() && !Arc.container().requestContext().isActive()) {
+                    Arc.container().requestContext().activate();
+                }
```
Comment (yrodiere) [inline extensions/security-jpa/runtime/src/main/java/io/quarkus/security/jpa/runtime/JpaIdentityProvider.java:44]: I'm not sure you're supposed to do this... In other areas we request that users explicitly add the `@ActivateRequestContext` annotation to their code if they need the request context.

If you still want to do this, you should deactivate the request context after you're done, no? See:

https://github.com/quarkusio/quarkus/blob/ebee0a4b66d1b4dba9c51165e78d22fbf3b90d3e/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ActivateRequestContextInterceptor.java#L121-L133

File: extensions/security-jpa/runtime/src/main/java/io/quarkus/security/jpa/runtime/JpaTrustedIdentityProvider.java [hunk 4]
```diff
@@ -37,21 +39,25 @@ public Uni<SecurityIdentity> authenticate(TrustedAuthenticationRequest request,
...
             @Override
             public SecurityIdentity get() {
-                EntityManager em = entityManagerFactory.createEntityManager();
-                ((org.hibernate.Session) em).setHibernateFlushMode(FlushMode.MANUAL);
-                ((org.hibernate.Session) em).setDefaultReadOnly(true);
-                try {
-                    return authenticate(em, request);
+                if (requireActiveCDIRequestContext() && !Arc.container().requestContext().isActive()) {
+                    Arc.container().requestContext().activate();
+                }
```
Comment (yrodiere) [inline extensions/security-jpa/runtime/src/main/java/io/quarkus/security/jpa/runtime/JpaTrustedIdentityProvider.java:44]: Same comment as above about activating the request context:

> I'm not sure you're supposed to do this... In other areas we request that users explicitly add the `@ActivateRequestContext` annotation to their code if they need the request context.
>
> If you still want to do this, you should deactivate the request context after you're done, no? See:
> 
> https://github.com/quarkusio/quarkus/blob/ebee0a4b66d1b4dba9c51165e78d22fbf3b90d3e/independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ActivateRequestContextInterceptor.java#L121-L133

File: extensions/security-jpa/runtime/src/main/java/io/quarkus/security/jpa/runtime/JpaIdentityProvider.java (hunk 3 above)
Comment (michalvavrik) [reply extensions/security-jpa/runtime/src/main/java/io/quarkus/security/jpa/runtime/JpaIdentityProvider.java:44]: > I'm not sure you're supposed to do this... In other areas we request that users explicitly add the @ActivateRequestContext annotation to their code if they need the request context.

I'm not happy to active request context, but users can't activate request context here, there is no place how they could do it on generated beans (and it wouldn't be on non-generated anyway unless they were supposed to provide their implementation).

> If you still want to do this, you should deactivate the request context after you're done, no? See:

Good point, thank you.

File: extensions/security-jpa/deployment/src/main/java/io/quarkus/security/jpa/deployment/QuarkusSecurityJpaProcessor.java (hunk 1 above)
Comment (michalvavrik) [reply extensions/security-jpa/deployment/src/main/java/io/quarkus/security/jpa/deployment/QuarkusSecurityJpaProcessor.java:268]: > Note it's technically possible for the default persistence unit to not have any configuration, in which case the map wouldn't return any entry.

ouch, I'd guess that, it is not optional.`public HibernateOrmConfigPersistenceUnit defaultPersistenceUnit;` How is that possible? Anyway, thank you

> Are you sure you want to rely on HibernateOrmConfig directly? It seems dodgy. Wouldn't it make more sense to consume io.quarkus.hibernate.orm.deployment.PersistenceUnitDescriptorBuildItem? It has the advantage of being present regardless of configuration quirks, and exposes the multitenancy strategy through getConfig().getMultiTenancyStrategy().

I definitely want to use descriptor, thank you, I'll change it.

Comment (yrodiere) [reply extensions/security-jpa/deployment/src/main/java/io/quarkus/security/jpa/deployment/QuarkusSecurityJpaProcessor.java:268]: > > Note it's technically possible for the default persistence unit to not have any configuration, in which case the map wouldn't return any entry.
> 
> ouch, I'd guess that, it is not optional.`public HibernateOrmConfigPersistenceUnit defaultPersistenceUnit;` How is that possible? Anyway, thank you

Hum, maybe I'm wrong... I'm sure I've seen this problem somewhere though. Anyway, even if it's not `null` right now, it definitely could be once we switch to `@ConfigMapping`. So, better stay away from this, and it seems you're working on it :)

File: extensions/security-jpa/deployment/src/test/resources/multitenant-persistence-unit/application.properties (hunk 2 above)
Comment (michalvavrik) [reply extensions/security-jpa/deployment/src/test/resources/multitenant-persistence-unit/application.properties]: That would mean I have to create user table manually or setup another PU. I've renamed it to `shared`. FYI verification that multitenancy works is done by dynamically setting `unknown` tenant which leads to error. I don't need different database for this.

**Review by yrodiere — APPROVED** [review]
I didn't look into pre-existing code, but at least the multi-tenancy integration seems to make sense. Thanks.

**Pull request merged**